Dz8 additional 2

angryQ opened the pull request.

angryQ committed:
initial

angryQ committed:
Добавила дополнительное требование к ДЗ8

Должно быть так:

Нажали кнопку 2
Появился контроллер 2.
Нажали кнопку 1.
Контроллер 1 появился над контроллером 2.
Должен сохраняться такой же порядок сверху вниз, как в массиве childs.

ios-review commented on DZ8-additional-2/DZ8-additional-2/ContainerViewController.swift:
```diff
@@ -0,0 +1,153 @@
+           let index = buttonsStackView.arrangedSubviews.firstIndex(of: sender)!
+           
+           if sender.title(for: .normal) == "Включить" {
+            addChildVC(childVC: childs[index], index: index)
+               sender.setTitle("Выключить", for: .normal)
+               if (defaultVC?.view.isDescendant(of: childVCStackView))! {
+                   removeChildVC(childVC: defaultVC!)
+               }
+           }
+           else {
+               removeChildVC(childVC: childs[index])
+               sender.setTitle("Включить", for: .normal)
```
Алгоритм не работает для 3 контент контроллеров.

Если раскомментировать в SceneDelegate создание purpleVC и добавить его в container, то поведение будет не ожидаемым.
"+" может быть падение (краш) приложения.

ios-review reviewed (changes requested):
Попробуете еще раз?

Если есть вопросы, буду рад ответить)

angryQ committed:
Починила отображение/скрытие 3-х вью

Не нравится используемый способ решения, но так хотя бы работает. Пока так. Подумаю над другим, может с использованием словаря, не знаю.

ios-review commented on DZ8-additional-2/DZ8-additional-2/ContainerViewController.swift:
```diff
@@ -9,145 +9,154 @@
+        countOfActiveViews = childs.count
+    }
+    
+    override func viewWillLayoutSubviews() {
+        super.viewWillLayoutSubviews()
+        
+        configureHolderStackView()
+        configureViewsStackView()
+    }
```
Данный метод может вызваться несколько раз за время жизни вью контроллера.
https://medium.com/good-morning-swift/ios-view-controller-life-cycle-2a0f02e74ff5

Рекомендую первоначальную настройку проводить в viewDidLoad.
Достаточно один раз сконфигурировать оба stack view.
viewDidLoad вызовется один раз.

ios-review commented on DZ8-additional-2/DZ8-additional-2/ContainerViewController.swift:
```diff
@@ -9,145 +9,154 @@
+        let index = buttonsStackView.arrangedSubviews.firstIndex(of: sender)!
+        
+        if sender.title(for: .normal) == "Включить" {
+            childs[index].view.isHidden = false
+            countOfActiveViews += 1
+            sender.setTitle("Выключить", for: .normal)
+            
+            guard let vc = defaultVC else {return}
+            if (vc.view.isDescendant(of: childVCStackView)) {
+                vc.willMove(toParent: nil)
+                vc.view.removeFromSuperview()
+                vc.removeFromParent()
+            }
+        }
+        else {
+            countOfActiveViews -= 1
+            removeChildVC(childVC: childs[index])
+            sender.setTitle("Включить", for: .normal)
+        }
```
Теперь алгоритм показа/скрытия дочернего контроллера изменился.

Задача контейнера управлять показом/скрытием дочерних контроллеров.
Эта задача сводится к следующему:
1) Добавить/удалить в/из иерархии контроллеров, чтобы дочерние контроллеры могли/перестали получать события жизненного цикла, вроде viewDidLoad/viewWillAppear/etc.
Если контроллер скрыт, то он не должен получать эти события.
Если скрыть его вью, но не удалить его из иерархии, то нарушится данный контракт.
Реализация дочернего контроллера может зависеть от методов жизненного цикла.
В реализации может выполняться очень затратная работа, которая не нужна, если контроллер скрыт.
2) Добавить/удалить в/из иерархии вью.

Сейчас реализация изменена так, что показывается/скрывается только вью у дочерних контроллеров, при этом сами они всегда остаются в иерархии контроллеров, так как были добавлены в нее с помощью метода addChildVC.

Вернете прошлую реализацию, сохранив при этом логику порядка отображения/скрытия?

ios-review reviewed (changes requested):
Алгоритм верный👍
Работает корректно для любого количества. Супер🔥

Теперь нужно вернуть добавлени/удаление контроллеров в/из иерархии при нажатию на кнопки.

angryQ committed:
Правки по рекомендациям Дмитрия

Починила удаление/ добавление вью с стэк

ios-review commented on DZ8-additional-2/DZ8-additional-2/ContainerViewController.swift:
```diff
@@ -78,43 +75,50 @@ class ContainerViewController: UIViewController {
             removeChildVC(childVC: childs[index])
+            removeIndexFromViewIndicesArray(index: index)
             sender.setTitle("Включить", for: .normal)
         }
```
Другой возможный алгоритм нахождения индекса вставки вью контроллера в стек вью.
1) Изменяем свойство isSelected у кнопок для понимания того, какие кнопки сейчас выделены.
Использование свойства состояния предпочтительнее, чем использование значение контента (текста) для определения собственно состояния)
`sender.isSelected.toggle()`
2) Вычисляем индекс следующим способом:
`let insertIndex = (buttonsStackView.arrangedSubviews as? [UIButton])?.filter { $0.isSelected }.firstIndex(of: sender) ?? 0`

ios-review reviewed (approved):
Контейнер работает ожидаемо и отлично!

На ваше усмотрение реализовать еще один вариант алгоритма нахождения индекса вставки вью контроллера.

ПР можно мержить.
